src/pymoca/tree.py

```diff
@@ -747,6 +775,494 @@ def _check_import_rules(
+        return instance
+
+    def _instantiate_class(
+        self,
+        orig_class: Union[ast.Class, ast.InstanceClass],
+        # Does InstanceElement need a fully_instantiated boolean flag?
+        modification_environment: ast.ClassModification,
+        parent: ast.Class,
+        partially=True,
+    ) -> ast.InstanceClass:
+        # Outline from spec 3.5 section 5.6.1 *Instantiation*:
+        # Definitions
+        #   - Element: Class, Component (Symbol in Pymoca), or Extends Clause
+        # 1. For element itself:
+        #   1. Create an instance of the class to be instantiated ("partially instantiated element")
+        #   2. Modifiers are merged for the element itself (but contained references are resolved during flattening)
+        #   3. Redeclare of element itself is done
+        # 2. For each element (Class or Component) in the local contents of the current element:
+        #   1. Apply step 1 to the element
+        #   2. Equations, algorithms, and annotations are copied into the component instance without merging
+        #      (but references in equations are resolved later during flattening)
+        # 3. For each element in the extends clauses of the current element:
+        #   1. Apply steps 1 and 2 to the element, replacing the extends clause with the extends instance
+        # 4. Lookup classes of extends and ensure it is identical to lookup result from step 3
+        # 5. Check that all children of the current element (including extends) with same name are identical
+        #    (error if not) and only keep one if so (to preserve function argument order)
+        # 6. Components are recursively instantiated
+
+        # 1.1. Partially instantiate the element itself and 1.2 merge modifiers
+        new_class = self._instantiate_partially(
+            orig_class,
+            modification_environment,
+            parent,
+        )
+
+        # 1.3. Redeclare of element itself is done
+        self._apply_class_redeclares(new_class, modification_environment)
+
+        # 2.1 Partially instantiate local classes and symbols
+        if isinstance(orig_class, ast.InstanceClass) and not orig_class.fully_instantiated:
+            from_class = new_class.ast_ref
+        else:
+            from_class = orig_class
+        for name, class_ in from_class.classes.items():
+            instance = self._instantiate_partially(
+                class_,
+                modification_environment,
+                new_class,
+            )
+            new_class.classes[name] = instance
+
+        for name, symbol in from_class.symbols.items():
+            instance = self._instantiate_partially(
+                symbol,
+                modification_environment,
+                new_class,
+            )
+            new_class.symbols[name] = instance
+
+        # 2.2 Copy local contents into the element itself
+        self._copy_class_contents(new_class, copy_extends=True, partially=partially)
```

Comment:
TODO: (Now) Improve the comments about partial instantiation and the `partially` flag
TODO: (Later) Look into refactoring partial instantiation

I think this is an area where there is leeway for implementations to innovate while still meeting the Modelica semantics. Any implementation needs to include imports to do proper name lookup, which is what partial instantiation is intended to enable. 

The way the code is written now allows imported names to be found earlier through the partial instance vs the restart of lookup from the beginning through `ast_ref` that it does if the name was not previously instantiated (see `tree.find_name()`). This implementation also keeps the imported name lookup uniform for `Class`, and both partially and fully instantiated `InstanceClass`.
 
The spec says:

> A partially instantiated class or component is an element that is ready to be instantiated; a partially instantiated element (i.e. class or component) is comprised of a reference to the original element (from the class tree) and the modifiers for that element (including a possible redeclaration).

That's what I tried to make `instantiate_partially` do. The `partially` flag came in when I realized `instantiate_extends` needed to do an `_instantiate_class`, but only partially. So I got a solution that works, but I can't say I love it. I did literally about 100 rebases to clean things up, but there is absolutely much room for improvement.

I have started on flattening and I know I'm going to run into this partially instantiated thing again so I appreciate your fresh perspective and comments!

Reply:
Building up a complete mental picture will take me some more time, but for now I'd say your approach here is good enough. What I wanted to say here mostly was that you seem to do an optimization of sorts (postpone something that's not needed yet), and that could warrant a comment saying that+why we're deviating from the example steps outlined in the spec.

Reply:
The `partially` flag has little/no effect on speed, so I decided to remove it in 1de4e69 and in the process simplify understanding of the code and eliminate this question about deviating from the spec. While it looks like I added `partially` as an optimization, it really was due to not understanding at the time exactly where partial instantiation left off and full instantiation began. I love this code a *little more* now. :heart_eyes: I think we also are avoiding some problems with this that would eventually trip us up in flattening.

Cudos @jackvreeken for spotting and prodding! 

Please take a look at the documentation and typing improvements I added to the instantiation API in 1de4e69 and if you agree please resolve the conversation or let me know what doesn't make sense. TIA.

Reply:
Most of it looks good to me, happy I could help :)

> Name lookup on the returned tree may still return an `ast.Class` or an `ast.InstanceClass` with its fully_instantiated` attribute set to `False`. If so, the class will need to be fully instantiated for flattening.

Was this always the case (before the this commit that removes the partially stuff)? If so, how would one go about "fully instantiating" a class? I feel like guaranteeing an `ast.InstanceClass` makes more sense, but I don't quite oversee how that would impact things.

The type of `instantiate` still says just `ast.InstanceClass` as well.

Reply:
How about this for the comment instead:

```
        The returned class will be fully instantiated, but name lookup
        in the instance tree may return `ast.Class` or
        `ast.InstanceClass` with the `fully_instantiated` flag set to
        `False`. If so, these cases will need to be fully instantiated
        for flattening by calling this method on the class.
```

In other words, some `InstanceClass` may have only been through `_partially_instantiate()`. This was always the case, even before this latest change. I'm saying this without having run into this yet in implementing flattening, but it the comment is consistent with the spec section 5.6.2 on flattening.

> The type of `instantiate` still says just `ast.InstanceClass` as well.

Isn't that what it should be? What am I missing?

Reply:
Ah, OK. I was a bit confused, because I thought your changes resulted in `instantiate`'s behavior changing. If it was always the case, then it is fine :) I also did not run into it myself, at least I don't remember seeing any scenario in where it would happen. I also did not specifically look for that yet though

Reply:
If you look at an instantiated class, you'll see contained classes that are not fully instantiated. As I look at this for `test_inheritance_instantiation`, I see some related issues however:

1. We are creating multiple instances (partially or fully instantiated) of the same class. (You mentioned this earlier.) This is unavoidable since there can be instances of the same class with different modifications, but there is likely room for improvement in reducing the number of instance copies and/or sharing the parts that are always the same. The spec mentions the possibility of reusing instantiated elements.
2. I wanted to try keeping the `InstanceTree` separate, but in hind sight I think we should try the alternate approach of just connecting into the class tree (which the spec mentions as a valid approach). Then we can quantify which approach is better. This will help reduce creation of unused partial instances.
3. Symbols are not getting marked as full instantiated in `_instantiate_symbol()`. This is somewhat trivial because it does not affect anything (symbols are always fully instantiated), but should be marked for completeness. I'll fix this.
4. We should think about the pros/cons of just using `Symbol` and its `class_modification` attribute for an instance (like previous Pymoca) vs `InstanceSymbol` and the separate `modification_environment`. I think the current use of `InstanceSymbol` may be a bit easier to understand, but then again it's maybe worth looking at the alternate approach. 
 
I wanted to note these things, but only plan to do item 3 for now.

Reply:
Added TODO in tree.py for items 1, 2, and 4 in cb617ce

Reply:
Fixed item 3 in 441741b